nebuchadnezzar/nebu/cli/assemble.py

```diff
@@ -70,26 +71,53 @@ def to_dom_resource_path(filename):
     return f"../resources/{filename}"
 
 
-def h5p_media_handler_factory(path_resolver: PathResolver, resource_dir: str):
-    @lru_cache
-    def handle_resource(
-        interactive_id: str, resource_orig_path: str, is_image: bool
+def media_handler_factory(
+    resource_dir: str, media_cache: dict[tuple, str] = {}
+):
+    def media_handler(
+        cache_key: tuple,
+        resource_abs_path: str | None,
+        is_image: bool,
     ):
-        resource_abs_path = path_resolver.find_interactives_path(
-            interactive_id, resource_orig_path
-        )
-        assert resource_abs_path is not None, \
-            f"Could not find resource: {interactive_id}:{resource_orig_path}"
-        sha1, metadata = get_media_metadata(resource_abs_path, is_image)
-        resource_dst = os.path.join(resource_dir, sha1)
-        shutil.move(resource_abs_path, resource_dst)
-        save_resource_metadata(metadata, resource_dir, sha1)
-        return to_dom_resource_path(sha1)
-
+        cached = media_cache.get(cache_key, None)
+        if cached is None:
+            assert resource_abs_path is not None, \
+                f"Missing resource: {cache_key}"
+            sha1, metadata = get_media_metadata(resource_abs_path, is_image)
+            resource_dst = os.path.join(resource_dir, sha1)
+            shutil.move(resource_abs_path, resource_dst)
+            save_resource_metadata(metadata, resource_dir, sha1)
+            cached = media_cache[cache_key] = to_dom_resource_path(sha1)
+        return cached
+    return media_handler
+
+
+def h5p_media_handler_factory(
+    path_resolver: PathResolver,
+    media_handler: Callable[[tuple, str | None, bool], str]
+):
     def h5p_media_handler(interactive_id, elem, uri_attrib, is_image):
-        resource_orig_path = elem.attrib[uri_attrib]
-        elem.attrib[uri_attrib] = handle_resource(
-            interactive_id, resource_orig_path, is_image
+        orig_path = elem.attrib[uri_attrib]
+        cache_key = (interactive_id, orig_path)
+        paths = path_resolver.find_interactives_paths(
+            interactive_id, orig_path
+        )
+        # Expect the public and private version of the file to be the same
+        # We would need to handle this problem even if we were copying all
+        # files into a temporary directory
+        if "public" in paths and "private" in paths:
+            private_path = paths["private"]
+            public_path = paths["public"]
+            private_checksums = get_checksums(private_path)
+            public_checksums = get_checksums(public_path)
+            assert private_checksums == public_checksums, (
+                "Files have the same name but different content:"
+                f"{public_path} and {private_path}"
+            )
+            os.unlink(private_path)
```

Comment:
Why not error if the file is in both public & private? Folks might incorrectly assume that if the file is in private then it's secret or be surprised that inside enki the private file disappeared (but is still in git).

I might be misunderstanding the code but I'm having trouble thinking of a use-case where we would want the same file to  be in both places. Also, it seems unexpected to delete the private file as part of the enki build